### PR TITLE
add tests

### PR DESCRIPTION
**Pull request by golddydev**
Added some tests

* PZ - resolved_addresses can't contain 'ada' (for `HANDLE` type)
* PZ - should deny if pz is disabled (for `NFT_SUBHANDLE` type)
* PZ - should pz if assignee signed virtual  (for `VIRTUAL_SUBHANDLE` type)
* REVOKE - should only revoke if public and expired
* REVOKE - should Deny revoke if public but NOT expired
* REVOKE - should Deny revoke if private but NOT signed by root
* UPDATE - should Update - extend if private & root_signed & good payment
* UPDATE - should Update - extend - if public & NOT root_signed && assignee signed with payment cred
* UPDATE - should Update - to_private - public & root_signed & expired & extended
* UPDATE - should Deny update if assignee NOT signed when public, and root handle not signed
* UPDATE - should Deny if we update pz rather than virtual & resolved address
* UPDATE - should Deny if we update nft rather than virtual & resolved address

Found an attack vector
* UPDATE - can Update - extend - without paying to main address and root address
`within_window` is attack vector
we can stop this by checking extended and within_window separately

Result

```bash
*success* - APPROVE - PERSONALIZE               'main - test most things' ( mem:12849641, cpu:6119632081, size:1510 )
*success* - APPROVE - PERSONALIZE               'unenforced defaults' ( mem:7333481, cpu:3403444998, size:1510 )
INFO (personalization:424:19) qr_inner_eye is not set correctly
*success* - DENY    - PERSONALIZE               'exclusives set, no creator' ( mem:0, cpu:0, size:1548 )
INFO (personalization:603:27) resolved_addresses can't contain 'ada'
*success* - DENY    - PERSONALIZE               'resolved_addresses can't contain ada' ( mem:0, cpu:0, size:1609 )
INFO (personalization:611:27) Root SubHandle settings prohibit Personalization
*success* - DENY    - PERSONALIZE               'pz is disabled' ( mem:0, cpu:0, size:1596 )
INFO (personalization:617:27) Tx not signed by virtual SubHandle holder
*success* - DENY    - PERSONALIZE               'should pz if assignee signed virtual' ( mem:0, cpu:0, size:1596 )
*success* - APPROVE - REVOKE                    'private mint' ( mem:501306, cpu:195537107, size:517 )
INFO (personalization:740:27) Publicly minted Virtual SubHandle hasn't expired
*success* - DENY    - REVOKE                    'private mint without root' ( mem:0, cpu:0, size:435 )
*success* - APPROVE - REVOKE                    'public and expired' ( mem:475339, cpu:182202799, size:395 )
INFO (personalization:740:27) Publicly minted Virtual SubHandle hasn't expired
*success* - DENY    - REVOKE                    'public mint not expired' ( mem:0, cpu:0, size:551 )
*success* - APPROVE - UPDATE                    'private mint address changed' ( mem:1521718, cpu:610130645, size:1488 )
*success* - APPROVE - UPDATE                    'private mint extended' ( mem:1529831, cpu:613159398, size:1482 )
*success* - APPROVE - UPDATE                    'public assignee signed' ( mem:1555903, cpu:622336417, size:1396 )
*success* - APPROVE - UPDATE                    'public assignee signed attack' ( mem:1318821, cpu:524595348, size:1296 )
*success* - APPROVE - UPDATE                    'public to private' ( mem:1540040, cpu:617902346, size:1482 )
INFO (personalization:795:30) No valid signature
*success* - DENY    - UPDATE                    'public assignee not signed' ( mem:0, cpu:0, size:1300 )
INFO (personalization:773:27) Restricted changes are not allowed
*success* - DENY    - UPDATE                    'update pz' ( mem:0, cpu:0, size:1526 )
INFO (personalization:773:27) Restricted changes are not allowed
*success* - DENY    - UPDATE                    'update nft' ( mem:0, cpu:0, size:1526 )
** SUMMARY **
   18 total tests
   18 successful
```

### CODE REVIEW

File: tests/fixtures.ts [hunk 1]
```diff
@@ -212,6 +234,10 @@ export class PzFixture extends Fixture {
     }
     
     async initialize(): Promise<PzFixture> {
+        /// in case `this.handleName` is updated after constructor
+        (this.oldCip68Datum.constructor_0[0] as any)['name'] = `$${this.handleName}`;
+        (this.newCip68Datum.constructor_0[0] as any)['name'] = `$${this.handleName}`;
```
Comment (papag00se) [inline tests/fixtures.ts]: I believe the constructor already does this?

File: tests/txTests.ts [hunk 2]
```diff
@@ -53,21 +62,145 @@ const runTests = async (file: string) => {
...
+        return await fixture.initialize();
+    }), false, "Restricted changes are not allowed"),
+
+    // should Deny update if assignee NOT signed when public, and root handle not signed
+    await tester.test("UPDATE", "assignee not signed", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        (fixture.oldCip68Datum.constructor_0[2] as any).virtual.public_mint = 1; /// make public
+        (fixture.newCip68Datum.constructor_0[2] as any).virtual.expires_slot = Date.now(); /// not extend
+        fixture.inputs?.splice(1, 0); /// remove 222 root_handle in inputs
+        fixture.outputs?.splice(0, 0); /// remove 222 root_handle in outputs
```
Comment (papag00se) [inline tests/txTests.ts]: In JavaScript, the second param in splice is the number of elements, yeah? So `0` does nothing?

File: tests/txTests.ts [hunk 3]
```diff
@@ -53,21 +62,145 @@ const runTests = async (file: string) => {
...
+        return await fixture.initialize();
+    }), false, "No valid signature"),
+
+    // should update if assignee signed when public, and root handle not signed
+    await tester.test("UPDATE", "assignee signed", new Test(program, async (hash) => {
+        const fixture = new UpdateFixture(hash);
+        (fixture.oldCip68Datum.constructor_0[2] as any).virtual.public_mint = 1; /// make public
+        (fixture.newCip68Datum.constructor_0[2] as any).virtual.expires_slot = Date.now(); /// not extend
+        fixture.inputs?.splice(1, 0); /// remove 222 root_handle in inputs (remove root_signed)
+        fixture.outputs?.splice(0, 0); /// remove 222 root_handle in outputs (remove root_signed)
```
Comment (papag00se) [inline tests/txTests.ts]: See above

File: tests/fixtures.ts [hunk 4]
```diff
@@ -617,7 +655,12 @@ export class UpdateFixture extends Fixture {
                 new helios.Value(BigInt(lovelace),
                     new helios.Assets([[POLICY_ID, [[`${AssetNameLabel.LBL_000}${Buffer.from(this.handleName).toString('hex')}`, 1]]]])),
                 helios.Datum.inline(helios.UplcData.fromCbor(this.newCip68DatumCbor))
-            )
+            ),
+            // new helios.TxOutput( // pay to main address
+            //     helios.Address.fromHash(helios.PubKeyHash.fromHex(this.adminSettings[5].slice(4))),
+            //     new helios.Value(BigInt(5000000)),
+            //     helios.Datum.inline(helios.UplcData.fromCbor(this.handleCbor))
+            // ),
```
Comment (papag00se) [inline tests/fixtures.ts]: I think this will be a required output for some of the tests. Is it added at the test level?

**Review by papag00se — CHANGES_REQUESTED** [review]
Love all the tests! Just a few comments. I had to push some changes to master, so pull the latest and update.

File: tests/txTests.ts (hunk 2 above)
Comment (golddydev) [reply tests/txTests.ts]: Whoops....
What am I thinking while typing that line? 🤣

File: tests/fixtures.ts (hunk 1 above)
Comment (golddydev) [reply tests/fixtures.ts]: This is when we change the handle name after constructor.
For example, when we test `PERSONALIZATION` with `NFT_SUBHANDLE` or `VIRTUAL_HANDLE`, we have to change handle name.

```typescript
fixture.handleName = 'dev@golddy'; /// nft subhandle
```

File: tests/fixtures.ts (hunk 4 above)
Comment (golddydev) [reply tests/fixtures.ts]: I was trying to add that output on some tests.
But didn't get it done yet.
I'll push once, I make it working.

File: tests/fixtures.ts (hunk 1 above)
Comment (golddydev) [reply tests/fixtures.ts]: resolved
Moved to certain test itself

**Comment by golddydev** [comment]
I have updated the comments on `top`

**Review by papag00se — APPROVED** [review]

**Pull request merged**